add very basic socket to l3roamd that reports the amount of connected…

christf opened the pull request:
have l3roamd answer with the client count on a socket.
Introduce -P <port> as command line switch

christf commented:
this fixes #2

lemoer commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+	char tmp[12];
+	memset(tmp, '\0', 12);
+	sprintf(tmp,"%zu", count);
+	write(fd, tmp, sizeof(tmp));
```
I think your process can crash here with `SIGPIPE`. For [reference](https://github.com/lemoer/bpfcountd/issues/1)

neocturne replied:
It's a socket, not a pipe.

lemoer replied:
Do I interpret this excerpt from `man 3 write` wrong?
```
EPIPE  A write was attempted on a socket  that  is  shut  down  for
       writing,  or  is no longer connected. In the latter case, if
       the socket is of type SOCK_STREAM, a  SIGPIPE  signal  shall
       also be sent to the thread.
```

Sounds like this also applies to sockets?

christf replied:
I think this is fixed now but was not able to test this yet with a malicious program creating the issue.

neocturne replied:
Ah, I was not aware of that. In that case, the SIGPIPE handler should be set to ignore the signal (this is always a good idea in daemons).

christf replied:
done

lemoer commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+
+	char tmp[12];
+	memset(tmp, '\0', 12);
+	sprintf(tmp,"%zu", count);
```
Maybe you should use `snprintf` instead. In this case, this should be okay for now, but if the code will be modified in future, it's often easier to see the limit. But I think this is contentious.

by the way: I think you missed a whitespace between `tmp,` and `"%zu"` :wink:

christf replied:
yup.

lemoer commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+void socket_init(socket_ctx *ctx) {
+	printf("init socket\n");
+	ctx->fd = socket(AF_INET6, SOCK_STREAM | SOCK_NONBLOCK, 0);
+	bind(ctx->fd, (struct sockaddr *)&ctx->socket, sizeof(ctx->socket));
```
What happens here, if the option `-P` was not given?

christf replied:
bad things happen. this should now be fixed for -P at least.

christf commented:
actually I believe using MSG_NOSIGNAL to be superior (ie actually working) to the hack before. [http://stackoverflow.com/questions/108183/how-to-prevent-sigpipes-or-handle-them-properly](url)

neocturne commented on src/main.c:
```diff
@@ -71,6 +72,11 @@ void loop(struct l3ctx *ctx) {
 	add_fd(efd, ctx->arp_ctx.fd, EPOLLIN);
 	add_fd(efd, ctx->intercom_ctx.fd, EPOLLIN | EPOLLET);
+
+	if (ctx->socket_ctx.fd >= 0) {
```
You check for >= 0 here, but you never set fd to -1 when it is unused.

christf replied:
done.

neocturne commented on src/socket.h:
```diff
@@ -0,0 +1,14 @@
+typedef struct {
+	struct l3ctx *l3ctx;
+	int fd;
+	struct sockaddr_in6 socket;
```
Let's make this a UNIX socket.

neocturne commented on src/main.c:
```diff
@@ -206,6 +216,12 @@ int main(int argc, char *argv[]) {
+				ctx.socket_ctx.socket.sin6_family = AF_INET6;
+				ctx.socket_ctx.socket.sin6_addr = in6addr_loopback;
+				ctx.socket_ctx.socket.sin6_port=htons(atoi(optarg));
+				printf("listening on port: %s\n", optarg);
```
Remove this message.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,49 @@
+void socket_handle_out(socket_ctx *ctx, size_t count) {
+	int fd  = accept(ctx->fd, (struct sockaddr*) NULL, NULL);
+
+	char tmp[BUFLEN];
```
No need to use a buffer here, just:
```C
dprintf(fd, "%zu", count);
```
Of course, as you can't use MSG_NOSIGNAL anymore, you'd also need to
```C
signal(SIGPIPE, SIG_IGN);
```
which is a good idea in any case.

I think for extensiblity, it would be nice to output some kind of JSON object instead of a plain number, similar to the fastd status socket. Maybe simply
```C
dprintf(fd, "{\"clients\":%zu}", count);
```
or something similar, depending on the way we plan to extend this format.

christf commented:
@NeoRaider this is in a working state now together with a respondd-module.
What else needs work in this PR?

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+void socket_init(socket_ctx *ctx, char *path) {
+	printf("initialize socket\n");
+	ctx->socket.sun_family = AF_UNIX;
+	strcpy(ctx->socket.sun_path, path);
```
Use strncpy.

christf replied:
done.

neocturne commented on src/main.c:
```diff
@@ -121,6 +129,7 @@ void usage() {
 	puts("  -a <ip6>          ip address of this node");
 	puts("  -c <file>         configuration file");
 	puts("  -p <prefix>       clientprefix"); // TODO mehrfache angabe sollte möglich sein
+	puts("  -P <socketpath>   provide statistics on this socket");
```
Maybe -S or -s would be better than -P?

christf replied:
-s it is.

neocturne commented on src/main.c:
```diff
@@ -173,11 +182,13 @@ int main(int argc, char *argv[]) {
 	ctx.arp_ctx.l3ctx = &ctx;
 	ctx.routemgr_ctx.l3ctx = &ctx;
+	ctx.socket_ctx.l3ctx = &ctx;
+	ctx.socket_ctx.fd =-1;
```
Add a space between = and -.

christf replied:
done.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+	ctx->socket.sun_family = AF_UNIX;
+	strcpy(ctx->socket.sun_path, path);
+	unlink(path);
+	signal(SIGPIPE, SIG_IGN);
```
I think this should be moved somewhere outside the socket initialization.

christf replied:
just the signal? where should it go?

neocturne replied:
Could be somewhere at the top of *loop*, or even earlier in *main*.

christf replied:
I placed it in main() for now. I am not sure though that is the sweet spot.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,47 @@
+}
+
+void socket_handle_out(socket_ctx *ctx, size_t count) {
+	int fd  = accept(ctx->fd, (struct sockaddr*) NULL, NULL);
```
The cast is not needed.

lemoer commented on src/socket.c:
```diff
@@ -32,16 +32,15 @@
 	printf("initialize socket\n");
 	ctx->socket.sun_family = AF_UNIX;
-	strcpy(ctx->socket.sun_path, path);
+	strncpy(ctx->socket.sun_path, path, strlen(path));
```
Well... That will ran into the same problem... `sun_path` is only 108 bytes long and you will produce an overflow if path is longer. Have a look at this example from `man 7 unix`:

``` c
struct sockaddr_un name;

// ...

memset(&name, 0, sizeof(struct sockaddr_un));

/* Bind socket to socket name. */

name.sun_family = AF_UNIX;
strncpy(name.sun_path, SOCKET_NAME,
        sizeof(name.sun_path) - 1);

// ...
```

You should also notice, that it's important to ensure the trailing `'\0'`, since `strncpy(dest, src, n)` does not ensure that, if there is no null byte amoge the first `n` bytes of `src`. In this example this is ensured by the `memset()`.

neocturne replied:
On Linux, this is actually not an issue (and l3roamd is very specific to Linux), as either a 0 termination or the *sizeof(struct sockaddr_un)* passed to bind work to limit the string length. So while we'd truncate a path when it is longer than sun_path, we wouldn't cause any memory issues.

Other possible approaches would be to output an error when the path doesn't fit into sizeof(name.sun_path) - 1 bytes, or to use a variable-size array for the sockaddr_un, which may include more space for sun_path (which is allowed on Linux and many other implementations).

neocturne replied:
For reference: In fastd, I use a variable-size sockaddr_un to allow arbitrary path lengths: https://git.universe-factory.net/fastd/tree/src/status.c#n244 (although I just noticed that my code is not completely correct either, as buf is not aligned to the minimum alignment of a sockaddr_un; I'll fix that soon...)

christf replied:
so what are we doing here? imo having a 108 char limit on the path of the socket is fine but I am open to copying the solution from fastd once that is reworked. :)

neocturne replied:
I've updated the fastd code, let's use that.

christf replied:
I have managed to copy & paste the solution I have to admit I am not sure if the approach is actually much better.

lemoer commented on src/socket.c:
```diff
@@ -32,16 +32,15 @@
-	signal(SIGPIPE, SIG_IGN);
 	ctx->fd = socket(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK, 0);
 	bind(ctx->fd, (struct sockaddr *)&ctx->socket, sizeof(ctx->socket));
 	listen(ctx->fd, 5);
```
The return values of `bind()` and `listen()` shouldn't be dropped silently. I would propose:

``` c
ret = bind(...);
if (ret == -1) {
    perror("bind");
    exit(EXIT_FAILURE);
}

ret = listen(...);
if (ret == -1) {
    perror("listen");
    exit(EXIT_FAILURE);
}
```

christf replied:
done.

neocturne commented on src/socket.h:
```diff
@@ -0,0 +1,15 @@
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <signal.h>
```
This include can now be moved to main.c.

christf replied:
done.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,46 @@
+	listen(ctx->fd, 5);
+}
+
+void socket_handle_out(socket_ctx *ctx, size_t count) {
```
The function should actually be called *socket_handle_in*, as POLLIN is used for listening sockets.

christf replied:
done.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,46 @@
+}
+
+void socket_handle_out(socket_ctx *ctx, size_t count) {
+	int fd  = accept(ctx->fd, NULL, NULL);
```
One space too much.

christf replied:
done.

neocturne commented on src/main.c:
```diff
@@ -206,6 +219,9 @@ int main(int argc, char *argv[]) {
 				break;
+			case 's':
+				socket_init(&ctx.socket_ctx, optarg);
+				break;;
```
One semicolon too much.

christf replied:
done.

neocturne commented on src/main.c:
```diff
@@ -206,6 +219,9 @@ int main(int argc, char *argv[]) {
 				ctx.clientmgr_ctx.export_table = atoi(optarg);
 				break;
+			case 's':
+				socket_init(&ctx.socket_ctx, optarg);
```
I don't like the way this initializes two sockets (leaking the first one) when -s is given multiple times. We should just save a pointer to the path in a variable and call *socket_init* with that after the getopt loop. I think setting fd the -1 should rather be done in *socket_init* when the given path is NULL.

christf replied:
done.

christf committed:
add very basic status-socket to l3roamd

The unix-socket will report a json structure that at the moment contains the amount of connected clients only.
This introduces a "-s" command line switch to activate the socket

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,75 @@
+	ctx->fd = socket(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK, 0);
+
+
+	int ret = bind(ctx->fd, (struct sockaddr *)&ctx->socket, sizeof(ctx->socket));
```
There are two bind calls now, this one must be removed (also remove the duplicate empty line).

christf replied:
done.

neocturne commented on src/socket.h:
```diff
@@ -0,0 +1,17 @@
+typedef struct {
+	struct l3ctx *l3ctx;
+	int fd;
+	struct sockaddr_un socket;
```
*socket* can be removed now from the struct.

christf replied:
done.

neocturne commented on src/main.c:
```diff
@@ -206,6 +219,9 @@ int main(int argc, char *argv[]) {
 				ctx.clientmgr_ctx.export_table = atoi(optarg);
 				break;
+			case 's':
+				socketpath=optarg;
```
Add spaces around the =

christf replied:
done.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,75 @@
+#include "error.h"
+
+void socket_init(socket_ctx *ctx, char *path) {
+	if (! path) {
```
Remove the space after the !

christf replied:
done.

neocturne commented on src/socket.c:
```diff
@@ -0,0 +1,75 @@
+		return;
+	}
+
+	printf("initialize socket\n");
```
This message is extremely vague, is should at least mention that this is the UNIX socket; adding the path would make it even clearer.

christf replied:
done.

christf committed:
fixup! add very basic status-socket to l3roamd

christf committed:
fixup! add very basic status-socket to l3roamd
